File: prototype/cogit/objects.py
"""Object schemas, validation, and preimage/ID computation.

Contract: docs/spec/object-format-v1.md. Unknown fields are rejected.
"""

import hashlib
import re

from .canonical import canonical_json_bytes
from .errors import UserError

OBJECT_TYPES = ("claim", "assertion", "mindset", "thought", "anchor", "annotation")

CLAIM_KINDS = (
    "user_preference",
    "tool_observation",
    "document_claim",
    "agent_decision",
    "policy_constraint",
)
ASSERTION_STATUSES = ("asserted", "refuted", "superseded")
SOURCE_TYPES = ("prompt", "tool", "file", "url", "system", "manual", "agent")
THOUGHT_OPERATIONS = ("commit", "merge", "checkout", "anchor", "import", "repair")

OID_RE = re.compile(r"^sha256:[0-9a-f]{64}$")
TIMESTAMP_RE = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$")
REF_SEGMENT_RE = re.compile(r"^[a-z0-9._-]+$")


def is_oid(value) -> bool:
    return isinstance(value, str) and bool(OID_RE.match(value))


def _require(obj, field, message=None):
    if field not in obj:
        raise UserError(message or f"object: missing required field '{field}'")
    return obj[field]


_MODE = __import__("threading").local()  # per-thread read/write validation mode


def _tolerant() -> bool:
    return getattr(_MODE, "tolerant", False)


def _check_keys(obj, allowed, optional=(), where="object"):
    if _tolerant():
        # ADR-0015: read mode tolerates unknown fields on known types —
        # an object written by a NEWER cogit is version skew, not corruption
        # (hash and known-field checks still apply; verify reports the skew)
        return
    unknown = set(obj) - set(allowed) - set(optional)
    if unknown:
        raise UserError(f"{where}: unknown fields rejected: {sorted(unknown)}")


def _check_str(value, field, nonempty=True):
    if not isinstance(value, str) or (nonempty and not value):
        raise UserError(f"object: field '{field}' must be a non-empty string")
    return value


def _check_oid(value, field):
    if not is_oid(value):
        raise UserError(f"object: field '{field}' must be a 'sha256:<64-hex>' reference")
    return value


def _check_timestamp(value, field):
    if not isinstance(value, str) or not TIMESTAMP_RE.match(value):
        raise UserError(f"object: field '{field}' must be ISO-8601 UTC with 'Z' suffix")
    return value


def _check_sorted_unique_oids(value, field):
    if not isinstance(value, list):
        raise UserError(f"object: field '{field}' must be a list")
    for item in value:
        _check_oid(item, field)
    if len(set(value)) != len(value):
        raise UserError(f"object: field '{field}' must not contain duplicates")
    if value != sorted(value):
        raise UserError(f"object: field '{field}' must be sorted lexicographically")
    return value


def _check_flat_object(value, field):
    """Qualifier/method-style maps: string keys, scalar values."""
    if not isinstance(value, dict):
        raise UserError(f"object: field '{field}' must be an object")
    for key, item in value.items():
        _check_str(key, f"{field} key")
        if not isinstance(item, (str, int, bool)) and item is not None:
            raise UserError(f"object: field '{field}.{key}' must be a scalar")
    return value


def _validate_claim(obj):
    _check_keys(
        obj,
        ("type", "kind", "subject", "predicate", "object", "qualifiers"),
        optional=("negates",),
        where="claim",
    )
    if _require(obj, "kind") not in CLAIM_KINDS:
        raise UserError(f"claim: kind must be one of {CLAIM_KINDS}")
    _check_str(_require(obj, "subject"), "subject")
    _check_str(_require(obj, "predicate"), "predicate")
    obj_value = _require(obj, "object")
    if not isinstance(obj_value, (str, int, bool)):
        raise UserError("claim: object must be a scalar (string, integer, or boolean)")
    _check_flat_object(_require(obj, "qualifiers"), "qualifiers")
    if "negates" in obj:
        _check_oid(obj["negates"], "negates")


def _validate_assertion(obj):
    _check_keys(
        obj,
        ("type", "claim", "status", "source", "confidence_bps", "asserted_at", "actor", "method"),
        optional=("premises",),
        where="assertion",
    )
    if "premises" in obj:
        premises = obj["premises"]
        if not isinstance(premises, list) or not premises:
            raise UserError("assertion: premises must be a non-empty array of assertion ids")
        for premise in premises:
            _check_oid(premise, "premises[]")
        if premises != sorted(set(premises)):
            raise UserError("assertion: premises must be sorted and unique")
    _check_oid(_require(obj, "claim"), "claim")
    if _require(obj, "status") not in ASSERTION_STATUSES:
        raise UserError(f"assertion: status must be one of {ASSERTION_STATUSES}")
    source = _require(obj, "source")
    if not isinstance(source, dict):
        raise UserError("assertion: source must be an object")
    _check_keys(source, ("type",), optional=("uri",), where="assertion.source")
    if _require(source, "type", "assertion.source: missing 'type'") not in SOURCE_TYPES:
        raise UserError(f"assertion: source.type must be one of {SOURCE_TYPES}")
    if "uri" in source:
        _check_str(source["uri"], "source.uri")
    confidence = _require(obj, "confidence_bps")
    if isinstance(confidence, bool) or not isinstance(confidence, int) or not 0 <= confidence <= 10000:
        raise UserError("assertion: confidence_bps must be an integer from 0 to 10000")
    _check_timestamp(_require(obj, "asserted_at"), "asserted_at")
    actor = _check_str(_require(obj, "actor"), "actor")
    if any(ch.isspace() for ch in actor):
        raise UserError("assertion: actor must not contain whitespace")
    method = _check_flat_object(_require(obj, "method"), "method")
    _check_str(method.get("type", ""), "method.type")


def _validate_mindset(obj):
    _check_keys(obj, ("type", "assertions", "created_at"), where="mindset")
    _check_sorted_unique_oids(_require(obj, "assertions"), "assertions")
    _check_timestamp(_require(obj, "created_at"), "created_at")


def _validate_thought(obj):
    _check_keys(
        obj,
        ("type", "parents", "mindset", "operation", "message", "author", "timestamp",
         "removals", "writer"),
        where="thought",
    )
    parents = _require(obj, "parents")
    if not isinstance(parents, list):
        raise UserError("thought: parents must be a list")
    for parent in parents:
        _check_oid(parent, "parents")
    if len(set(parents)) != len(parents):
        raise UserError("thought: parents must not contain duplicates")
    _check_oid(_require(obj, "mindset"), "mindset")
    if _require(obj, "operation") not in THOUGHT_OPERATIONS:
        raise UserError(f"thought: operation must be one of {THOUGHT_OPERATIONS}")
    _check_str(_require(obj, "message"), "message")
    _check_str(_require(obj, "author"), "author")
    _check_timestamp(_require(obj, "timestamp"), "timestamp")
    if "removals" in obj:
        # ADR-0014: durable removal provenance — optional, additive
        removals = obj["removals"]
        if not isinstance(removals, list) or not removals:
            raise UserError("thought: removals must be a non-empty list when present")
        seen = []
        for entry in removals:
            if not isinstance(entry, dict) or set(entry) != {"assertion", "reason"}:
                raise UserError("thought: each removal needs exactly {assertion, reason}")
            _check_oid(entry["assertion"], "removals.assertion")
            reason = entry["reason"]
            if not isinstance(reason, str) or not reason:
                raise UserError("thought: removal reason must be a non-empty string")
            seen.append(entry["assertion"])
        if seen != sorted(seen) or len(set(seen)) != len(seen):
            raise UserError("thought: removals must be sorted by assertion and unique")
    if "writer" in obj:
        # ADR-0016: writer provenance — optional, additive
        writer = obj["writer"]
        if not isinstance(writer, str) or not 0 < len(writer) <= 64:
            raise UserError("thought: writer must be a non-empty string of at most 64 chars")
        impl_part, sep, version_part = writer.partition("/")
        if not sep or not impl_part or not version_part or "/" in version_part \
                or any(ch.isspace() or not ch.isprintable() for ch in writer):
            raise UserError("thought: writer must be a single '<impl>/<version>' token")


def _validate_anchor(obj):
    _check_keys(obj, ("type", "name", "target", "created_at", "author"), where="anchor")
    name = _check_str(_require(obj, "name"), "name")
    if not REF_SEGMENT_RE.match(name):
        raise UserError("anchor: name must be a valid ref segment")
    _check_oid(_require(obj, "target"), "target")
    _check_timestamp(_require(obj, "created_at"), "created_at")
    _check_str(_require(obj, "author"), "author")


def _validate_annotation(obj):
    _check_keys(
        obj,
        ("type", "target", "namespace", "body", "author", "created_at", "parents"),
        where="annotation",
    )
    _check_oid(_require(obj, "target"), "target")
    namespace = _check_str(_require(obj, "namespace"), "namespace")
    if not REF_SEGMENT_RE.match(namespace):
        raise UserError("annotation: namespace must be a valid ref segment")
    _check_str(_require(obj, "body"), "body")
    _check_str(_require(obj, "author"), "author")
    _check_timestamp(_require(obj, "created_at"), "created_at")
    parents = _require(obj, "parents")
    if not isinstance(parents, list):
        raise UserError("annotation: parents must be a list")
    for parent in parents:
        _check_oid(parent, "parents")
    if len(set(parents)) != len(parents):
        raise UserError("annotation: parents must not contain duplicates")


_VALIDATORS = {
    "claim": _validate_claim,
    "assertion": _validate_assertion,
    "mindset": _validate_mindset,
    "thought": _validate_thought,
    "anchor": _validate_anchor,
    "annotation": _validate_annotation,
}


def validate_object(obj, mode="write") -> str:
    """Validate an object against its schema; return the object type.

    mode="write" (default) is STRICT: unknown fields are rejected — we only
    ever write what we understand. mode="read" is tolerant of unknown fields
    on known types (ADR-0015): additive format extensions must not turn
    older readers' view of newer objects into CorruptionError; integrity is
    still guaranteed by the hash and the known-field checks.
    """
    if not isinstance(obj, dict):
        raise UserError("object: must be a JSON object")
    obj_type = obj.get("type")
    if obj_type not in OBJECT_TYPES:
        raise UserError(f"object: type must be one of {OBJECT_TYPES}")
    _MODE.tolerant = mode == "read"
    try:
        _VALIDATORS[obj_type](obj)
    finally:
        _MODE.tolerant = False
    return obj_type


def encode_object(obj):
    """Return (object_id, preimage_bytes) for a validated object."""
    obj_type = validate_object(obj)
    body = canonical_json_bytes(obj)
    preimage = f"{obj_type} {len(body)}".encode("ascii") + b"\x00" + body
    oid = "sha256:" + hashlib.sha256(preimage).hexdigest()
    return oid, preimage

File: crates/cogit-core/src/objects.rs
//! Object schemas, validation, and preimage/ID computation.
//! Port of the reference validators (prototype/cogit/objects.py); unknown
//! fields are rejected.

use serde_json::{Map, Value};
use sha2::{Digest, Sha256};

use crate::canonical::canonical_json_bytes;
use crate::error::{CoreError, Result};

pub const OBJECT_TYPES: [&str; 6] = [
    "claim",
    "assertion",
    "mindset",
    "thought",
    "anchor",
    "annotation",
];

const CLAIM_KINDS: [&str; 5] = [
    "user_preference",
    "tool_observation",
    "document_claim",
    "agent_decision",
    "policy_constraint",
];
const ASSERTION_STATUSES: [&str; 3] = ["asserted", "refuted", "superseded"];
const SOURCE_TYPES: [&str; 7] = ["prompt", "tool", "file", "url", "system", "manual", "agent"];
const THOUGHT_OPERATIONS: [&str; 6] = ["commit", "merge", "checkout", "anchor", "import", "repair"];

fn user_err<T>(msg: impl Into<String>) -> Result<T> {
    Err(CoreError::User(msg.into()))
}

pub fn is_oid(value: &str) -> bool {
    value.len() == 71
        && value.starts_with("sha256:")
        && value[7..].bytes().all(|b| b.is_ascii_hexdigit() && !b.is_ascii_uppercase())
}

fn is_timestamp(value: &str) -> bool {
    let bytes = value.as_bytes();
    if bytes.len() != 20 {
        return false;
    }
    for (i, b) in bytes.iter().enumerate() {
        let ok = match i {
            4 | 7 => *b == b'-',
            10 => *b == b'T',
            13 | 16 => *b == b':',
            19 => *b == b'Z',
            _ => b.is_ascii_digit(),
        };
        if !ok {
            return false;
        }
    }
    true
}

fn is_ref_segment(value: &str) -> bool {
    !value.is_empty()
        && value
            .bytes()
            .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'.' || b == b'_' || b == b'-')
}

fn check_keys(map: &Map<String, Value>, allowed: &[&str], optional: &[&str], where_: &str) -> Result<()> {
    for key in map.keys() {
        if !allowed.contains(&key.as_str()) && !optional.contains(&key.as_str()) {
            return user_err(format!("{where_}: unknown fields rejected: {key}"));
        }
    }
    Ok(())
}

fn get_str<'a>(map: &'a Map<String, Value>, field: &str, where_: &str) -> Result<&'a str> {
    match map.get(field).and_then(Value::as_str) {
        Some(s) if !s.is_empty() => Ok(s),
        _ => user_err(format!("{where_}: field '{field}' must be a non-empty string")),
    }
}

fn get_oid<'a>(map: &'a Map<String, Value>, field: &str, where_: &str) -> Result<&'a str> {
    let value = get_str(map, field, where_)?;
    if !is_oid(value) {
        return user_err(format!("{where_}: field '{field}' must be a 'sha256:<64-hex>' reference"));
    }
    Ok(value)
}

fn get_timestamp(map: &Map<String, Value>, field: &str, where_: &str) -> Result<()> {
    let value = get_str(map, field, where_)?;
    if !is_timestamp(value) {
        return user_err(format!("{where_}: field '{field}' must be ISO-8601 UTC with 'Z' suffix"));
    }
    Ok(())
}

fn get_oid_list<'a>(
    map: &'a Map<String, Value>,
    field: &str,
    require_sorted: bool,
    where_: &str,
) -> Result<Vec<&'a str>> {
    let items = match map.get(field).and_then(Value::as_array) {
        Some(items) => items,
        None => return user_err(format!("{where_}: field '{field}' must be a list")),
    };
    let mut out = Vec::with_capacity(items.len());
    for item in items {
        match item.as_str() {
            Some(s) if is_oid(s) => out.push(s),
            _ => return user_err(format!("{where_}: field '{field}' must contain object ids")),
        }
    }
    let mut dedup = out.clone();
    dedup.sort_unstable();
    dedup.dedup();
    if dedup.len() != out.len() {
        return user_err(format!("{where_}: field '{field}' must not contain duplicates"));
    }
    if require_sorted && dedup != out {
        return user_err(format!("{where_}: field '{field}' must be sorted lexicographically"));
    }
    Ok(out)
}

fn check_flat_object<'a>(map: &'a Map<String, Value>, field: &str, where_: &str) -> Result<&'a Map<String, Value>> {
    let inner = match map.get(field).and_then(Value::as_object) {
        Some(inner) => inner,
        None => return user_err(format!("{where_}: field '{field}' must be an object")),
    };
    for (key, item) in inner {
        let scalar = matches!(item, Value::Null | Value::Bool(_) | Value::String(_))
            || matches!(item, Value::Number(n) if !n.is_f64());
        if !scalar {
            return user_err(format!("{where_}: field '{field}.{key}' must be a scalar"));
        }
    }
    Ok(inner)
}

fn validate_claim(map: &Map<String, Value>) -> Result<()> {
    check_keys(map, &["type", "kind", "subject", "predicate", "object", "qualifiers"], &["negates"], "claim")?;
    let kind = get_str(map, "kind", "claim")?;
    if !CLAIM_KINDS.contains(&kind) {
        return user_err(format!("claim: kind must be one of {CLAIM_KINDS:?}"));
    }
    get_str(map, "subject", "claim")?;
    get_str(map, "predicate", "claim")?;
    let object_ok = match map.get("object") {
        Some(Value::String(_)) | Some(Value::Bool(_)) => true,
        Some(Value::Number(n)) => !n.is_f64(),
        _ => false,
    };
    if !object_ok {
        return user_err("claim: object must be a scalar (string, integer, or boolean)");
    }
    check_flat_object(map, "qualifiers", "claim")?;
    if map.contains_key("negates") {
        get_oid(map, "negates", "claim")?;
    }
    Ok(())
}

fn validate_assertion(map: &Map<String, Value>) -> Result<()> {
    check_keys(
        map,
        &["type", "claim", "status", "source", "confidence_bps", "asserted_at", "actor", "method"],
        &["premises"],
        "assertion",
    )?;
    if map.contains_key("premises") {
        let premises = get_oid_list(map, "premises", true, "assertion")?;
        if premises.is_empty() {
            return user_err("assertion: premises must be a non-empty array of assertion ids");
        }
    }
    get_oid(map, "claim", "assertion")?;
    let status = get_str(map, "status", "assertion")?;
    if !ASSERTION_STATUSES.contains(&status) {
        return user_err(format!("assertion: status must be one of {ASSERTION_STATUSES:?}"));
    }
    let source = match map.get("source").and_then(Value::as_object) {
        Some(source) => source,
        None => return user_err("assertion: source must be an object"),
    };
    check_keys(source, &["type"], &["uri"], "assertion.source")?;
    let source_type = get_str(source, "type", "assertion.source")?;
    if !SOURCE_TYPES.contains(&source_type) {
        return user_err(format!("assertion: source.type must be one of {SOURCE_TYPES:?}"));
    }
    if source.contains_key("uri") {
        get_str(source, "uri", "assertion.source")?;
    }
    let confidence = map.get("confidence_bps").and_then(Value::as_i64);
    match confidence {
        Some(c) if (0..=10000).contains(&c) && !map["confidence_bps"].is_boolean() => {}
        _ => return user_err("assertion: confidence_bps must be an integer from 0 to 10000"),
    }
    get_timestamp(map, "asserted_at", "assertion")?;
    let actor = get_str(map, "actor", "assertion")?;
    if actor.chars().any(char::is_whitespace) {
        return user_err("assertion: actor must not contain whitespace");
    }
    let method = check_flat_object(map, "method", "assertion")?;
    match method.get("type").and_then(Value::as_str) {
        Some(t) if !t.is_empty() => Ok(()),
        _ => user_err("assertion: method.type must be a non-empty string"),
    }
}

fn validate_mindset(map: &Map<String, Value>) -> Result<()> {
    check_keys(map, &["type", "assertions", "created_at"], &[], "mindset")?;
    get_oid_list(map, "assertions", true, "mindset")?;
    get_timestamp(map, "created_at", "mindset")
}

fn validate_thought(map: &Map<String, Value>) -> Result<()> {
    check_keys(
        map,
        &["type", "parents", "mindset", "operation", "message", "author", "timestamp"],
        &["removals"],
        "thought",
    )?;
    get_oid_list(map, "parents", false, "thought")?; // semantic order (CQ-006)
    get_oid(map, "mindset", "thought")?;
    let operation = get_str(map, "operation", "thought")?;
    if !THOUGHT_OPERATIONS.contains(&operation) {
        return user_err(format!("thought: operation must be one of {THOUGHT_OPERATIONS:?}"));
    }
    get_str(map, "message", "thought")?;
    get_str(map, "author", "thought")?;
    get_timestamp(map, "timestamp", "thought")?;
    if let Some(removals) = map.get("removals") {
        // ADR-0014: durable removal provenance — optional, additive
        let entries = removals
            .as_array()
            .filter(|list| !list.is_empty())
            .ok_or_else(|| {
                CoreError::User("thought: removals must be a non-empty list when present".into())
            })?;
        let mut seen: Vec<String> = Vec::new();
        for entry in entries {
            let entry_map = entry.as_object().ok_or_else(|| {
                CoreError::User("thought: each removal needs exactly {assertion, reason}".into())
            })?;
            if entry_map.len() != 2
                || !entry_map.contains_key("assertion")
                || !entry_map.contains_key("reason")
            {
                return user_err("thought: each removal needs exactly {assertion, reason}");
            }
            let assertion = get_oid(entry_map, "assertion", "thought.removals")?;
            let reason = entry_map.get("reason").and_then(Value::as_str).unwrap_or("");
            if reason.is_empty() {
                return user_err("thought: removal reason must be a non-empty string");
            }
            seen.push(assertion.to_owned());
        }
        let mut sorted = seen.clone();
        sorted.sort();
        sorted.dedup();
        if seen != sorted {
            return user_err("thought: removals must be sorted by assertion and unique");
        }
    }
    Ok(())
}

fn validate_anchor(map: &Map<String, Value>) -> Result<()> {
    check_keys(map, &["type", "name", "target", "created_at", "author"], &[], "anchor")?;
    let name = get_str(map, "name", "anchor")?;
    if !is_ref_segment(name) {
        return user_err("anchor: name must be a valid ref segment");
    }
    get_oid(map, "target", "anchor")?;
    get_timestamp(map, "created_at", "anchor")?;
    get_str(map, "author", "anchor")?;
    Ok(())
}

fn validate_annotation(map: &Map<String, Value>) -> Result<()> {
    check_keys(
        map,
        &["type", "target", "namespace", "body", "author", "created_at", "parents"],
        &[],
        "annotation",
    )?;
    get_oid(map, "target", "annotation")?;
    let namespace = get_str(map, "namespace", "annotation")?;
    if !is_ref_segment(namespace) {
        return user_err("annotation: namespace must be a valid ref segment");
    }
    get_str(map, "body", "annotation")?;
    get_str(map, "author", "annotation")?;
    get_timestamp(map, "created_at", "annotation")?;
    get_oid_list(map, "parents", false, "annotation")?; // chain order
    Ok(())
}

/// Validate an object against its schema; return the object type.
pub fn validate_object(value: &Value) -> Result<&str> {
    let map = match value.as_object() {
        Some(map) => map,
        None => return user_err("object: must be a JSON object"),
    };
    let obj_type = match map.get("type").and_then(Value::as_str) {
        Some(t) if OBJECT_TYPES.contains(&t) => t,
        _ => return user_err(format!("object: type must be one of {OBJECT_TYPES:?}")),
    };
    match obj_type {
        "claim" => validate_claim(map)?,
        "assertion" => validate_assertion(map)?,
        "mindset" => validate_mindset(map)?,
        "thought" => validate_thought(map)?,
        "anchor" => validate_anchor(map)?,
        "annotation" => validate_annotation(map)?,
        _ => unreachable!(),
    }
    Ok(obj_type)
}

/// (object_id, preimage bytes) for a validated object.
pub fn encode_object(value: &Value) -> Result<(String, Vec<u8>)> {
    let obj_type = validate_object(value)?.to_owned();
    let body = canonical_json_bytes(value)?;
    let mut preimage = format!("{} {}", obj_type, body.len()).into_bytes();
    preimage.push(0);
    preimage.extend_from_slice(&body);
    let digest = Sha256::digest(&preimage);
    let mut oid = String::with_capacity(71);
    oid.push_str("sha256:");
    for byte in digest {
        oid.push_str(&format!("{byte:02x}"));
    }
    Ok((oid, preimage))
}
